src/serialization/sb2.js

```diff
@@ -470,12 +470,21 @@ const parseScratchObject = function (object, runtime, extensions, topLevel, zip)
     if (object.hasOwnProperty('variables')) {
         for (let j = 0; j < object.variables.length; j++) {
             const variable = object.variables[j];
+            // A variable is a cloud variable if:
+            // - the project says it's a cloud variable, and
+            // - it's a stage variable, and
+            // - the runtime can support another cloud variable
+            const isCloud = variable.isPersistent && topLevel &&
+                // It's important that this part of the check goes last
+                // because it will update the cloud variable limit counter.
+                runtime.canAddNewCloudVariable();
             const newVariable = new Variable(
                 getVariableId(variable.name, Variable.SCALAR_TYPE),
                 variable.name,
                 Variable.SCALAR_TYPE,
-                variable.isPersistent
+                isCloud
             );
+            if (isCloud && !runtime.hasCloudData) runtime.hasCloudData = true;
```

Comment:
I ended up deciding to just use this check (here and in sb3) to set the runtime `hasCloudData` flag since it is more reliable and @thisandagain said that we're not using the existing sb2 flag for research.